### PR TITLE
add paw3204 mouse sensor

### PR DESCRIPTION
**Pull request by gompa**
add paw 3204 mouse sensors driver 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x ] My change requires a change to the documentation.
- [ x] I have updated the documentation accordingly.
- [x ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/pointing_device.h [hunk 1]
```diff
@@ -20,8 +20,9 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include <stdint.h>
 #include "host.h"
 #include "report.h"
-
-#if defined(POINTING_DEVICE_DRIVER_adns5050)
+#if defined(POINTING_DEVICE_DRIVER_paw3204)
```
Comment (drashna) [inline quantum/pointing_device.h:23]: Could you move this to lower in the list. Would like to keep this alphabetical.

File: quantum/pointing_device_drivers.c [hunk 2]
```diff
@@ -26,7 +26,30 @@
 #define constrain_hid(amt) ((amt) < -127 ? -127 : ((amt) > 127 ? 127 : (amt)))
 
 // get_report functions should probably be moved to their respective drivers.
-#if defined(POINTING_DEVICE_DRIVER_adns5050)
+
+#if defined(POINTING_DEVICE_DRIVER_paw3204)
```
Comment (drashna) [inline quantum/pointing_device_drivers.c:30]: same here.

File: drivers/sensors/paw3204.c [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        else
+            writePinLow(PAW3204_DATA);
+        // wait_us(2);
+        writePinHigh(PAW3204_SCLK);
+    }
+
+    wait_us(4);
+}
+
+int8_t convert_twoscomp(uint8_t data) {
```
Comment (drashna) [inline drivers/sensors/paw3204.c:86]: is the value for the sensor only able to return -127 - 127? 

Eg, if it's capable of a higher value, then I would prefer that this is grabbed and handed over, and then let the pointing_device_driver.c code handle constraining the limit. 

This conforms better to the existing behavior, and makes #16371 much more "Drop in" for this sensor.

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — COMMENTED** [review]
Also, could you add a set/get cpi function that sets the resolution

File: drivers/sensors/paw3204.c (hunk 3 above)
Comment (gompa) [reply drivers/sensors/paw3204.c:86]: according to the data sheet the report range is  –128 ~ +127. what would the best way of doing this be ?

Comment (drashna) [reply drivers/sensors/paw3204.c:86]: then, as is is fine.  It just means that it won't benefit from the linked PR, and that's fine.  my concern was if the sensor was capable of it, at all (some of the other ones are).

File: docs/feature_pointing_device.md [hunk 4]
```diff
@@ -12,6 +12,31 @@ POINTING_DEVICE_ENABLE = yes
 
 There are a number of sensors that are supported by default. Note that only one sensor can be enabled by `POINTING_DEVICE_DRIVER` at a time.  If you need to enable more than one sensor, then you need to implement it manually.
 
+### PAW 3204 Sensor
```
Comment (drashna) [inline docs/feature_pointing_device.md:15]: This should be moved too

**Comment by gompa** [comment]
> Also, could you add a set/get cpi function that sets the resolution

something like this  ?

File: drivers/sensors/paw3204.c [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    wait_us(5);
+
+    uint8_t byte = PAW3204_serial_read();
+
+    return byte;
+}
+
+void PAW3204_set_cpi(uint16_t cpi) {
+    uint8_t cpival = constrain(cpi, 0x0, 0x6); // limits to 0--119
```
Comment (drashna) [inline drivers/sensors/paw3204.c:121]: To properly handle 400-1600, i think you need: 
```suggestion
    uint8_t cpival = constrain((cpi / 200) - 2, 0x0, 0x6);
```

File: drivers/sensors/paw3204.c [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    PAW3204_write_reg(REG_SETUP, cpival);
+}
+
+uint16_t PAW3204_get_cpi(void) {
+    uint8_t cpival = PAW3204_read_reg(REG_SETUP);
+    return (uint16_t)(cpival);
```
Comment (drashna) [inline drivers/sensors/paw3204.c:127]: And to get the proper value back: 
```suggestion
    return (uint16_t)((cpival + 2) & 0x06 * 200;
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: drivers/sensors/paw3204.c (hunk 6 above)
Comment (gompa) [reply drivers/sensors/paw3204.c]: i that gave a 0 as return seems to be fixed with the new commit

**Review by drashna — APPROVED** [review]

File: docs/feature_pointing_device.md [hunk 7]
```diff
@@ -113,6 +114,30 @@ This supports the Cirque Pinnacle 1CA027 Touch Controller, which is used in the
 |`CIRQUE_PINNACLE_SPI_CS_PIN`   | (Required) Sets the Cable Select pin connected to the sensor.          | _not defined_ |
 
 Default Scaling/CPI is 1024.
+### PAW 3204 Sensor
```
Comment (fauxpark) [inline docs/feature_pointing_device.md:117]:
```suggestion

### PAW 3204 Sensor
```

File: docs/feature_pointing_device.md [hunk 8]
```diff
@@ -12,6 +12,7 @@ POINTING_DEVICE_ENABLE = yes
 
 There are a number of sensors that are supported by default. Note that only one sensor can be enabled by `POINTING_DEVICE_DRIVER` at a time.  If you need to enable more than one sensor, then you need to implement it manually.
 
+
```
Comment (fauxpark) [inline docs/feature_pointing_device.md:15]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Pull request closed without merging**